### ISSUE TITLE
compaction task hangs without error

### ISSUE THREAD

**Issue by hzxa21** [risingwavelabs/risingwave#15209]
Recently in our testing pipeline, we found that some compaction tasks hang for a long time (>45min) without any error. Although the meta-side task expiration mechanism can make sure task will be forced cancelled if no progress has been made for some time, it is worth investigating why the task can hang at the first place.

![image](https://github.com/risingwavelabs/risingwave/assets/5518566/4b014556-1139-4434-89b0-765e1f6c467e)

Note that the recent occurrences are all in the test run when testing switching to opendal s3:
1. [nexmark-kafka-benchmark-20240221-075859](https://grafana.test.risingwave-cloud.xyz/explore?panes=%7B%22Y3S%22:%7B%22datasource%22:%22PE59595AED52CF917%22,%22queries%22:%5B%7B%22refId%22:%22A%22,%22expr%22:%22%7Bnamespace%3D%5C%22nexmark-kafka-benchmark-20240221-075859%5C%22,%20component%3D%5C%22meta%5C%22%7D%20%7C%3D%20%60too%20long%60%22,%22queryType%22:%22range%22,%22datasource%22:%7B%22type%22:%22loki%22,%22uid%22:%22PE59595AED52CF917%22%7D,%22editorMode%22:%22builder%22%7D%5D,%22range%22:%7B%22from%22:%221708445018000%22,%22to%22:%221708518818000%22%7D%7D%7D&schemaVersion=1&orgId=1)
2. [nexmark-kafka-benchmark-20240220-132426](https://grafana.test.risingwave-cloud.xyz/explore?orgId=1&panes=%7B%22Y3S%22:%7B%22datasource%22:%22PE59595AED52CF917%22,%22queries%22:%5B%7B%22refId%22:%22A%22,%22expr%22:%22%7Bnamespace%3D%5C%22nexmark-kafka-benchmark-20240220-132426%5C%22,%20pod%3D%5C%22benchmark-risingwave-meta-m-0%5C%22%7D%20%7C~%20%60%28%3Fi%29too%20long%60%22,%22queryType%22:%22range%22,%22datasource%22:%7B%22type%22:%22loki%22,%22uid%22:%22PE59595AED52CF917%22%7D,%22editorMode%22:%22builder%22%7D%5D,%22range%22:%7B%22from%22:%221708453320000%22,%22to%22:%221708458932000%22%7D%7D%7D&schemaVersion=1)
3. [nexmark-kafka-benchmark-20240220-132306](https://grafana.test.risingwave-cloud.xyz/explore?panes=%7B%22Y3S%22:%7B%22datasource%22:%22PE59595AED52CF917%22,%22queries%22:%5B%7B%22refId%22:%22A%22,%22expr%22:%22%7Bnamespace%3D%5C%22nexmark-kafka-benchmark-20240220-132306%5C%22,%20component%3D%5C%22meta%5C%22%7D%20%7C%3D%20%60too%20long%60%22,%22queryType%22:%22range%22,%22datasource%22:%7B%22type%22:%22loki%22,%22uid%22:%22PE59595AED52CF917%22%7D,%22editorMode%22:%22builder%22%7D%5D,%22range%22:%7B%22from%22:%221708445018000%22,%22to%22:%221708518818000%22%7D%7D%7D&schemaVersion=1&orgId=1)

**Comment by hzxa21** [comment]
cc @Li0k @wcy-fdu

**Comment by st1page** [comment]
This not only happen with OpenDAL s3
https://github.com/risingwavelabs/risingwave/issues/15169
https://buildkite.com/risingwave-test/nexmark-benchmark/builds/3092

q17: https://grafana.test.risingwave-cloud.xyz/d/EpkBw5W4k/risingwave-dev-dashboard?orgId=1&var-datasource=Prometheus:%20test-useast1-eks-a&from=1708450421000&to=1708452226000&var-namespace=nexmark-bs-15-105-daily-20240220

q19: https://grafana.test.risingwave-cloud.xyz/d/EpkBw5W4k/risingwave-dev-dashboard?orgId=1&var-datasource=Prometheus:%20test-useast1-eks-a&from=1708454470000&to=1708456275000&var-namespace=nexmark-bs-15-105-daily-20240220

**Comment by hzxa21** [comment]
> This not only happen with OpenDAL s3 #15169 https://buildkite.com/risingwave-test/nexmark-benchmark/builds/3092
> 
> q17: https://grafana.test.risingwave-cloud.xyz/d/EpkBw5W4k/risingwave-dev-dashboard?orgId=1&var-datasource=Prometheus:%20test-useast1-eks-a&from=1708450421000&to=1708452226000&var-namespace=nexmark-bs-15-105-daily-20240220
> 
> q19: https://grafana.test.risingwave-cloud.xyz/d/EpkBw5W4k/risingwave-dev-dashboard?orgId=1&var-datasource=Prometheus:%20test-useast1-eks-a&from=1708454470000&to=1708456275000&var-namespace=nexmark-bs-15-105-daily-20240220

Search for the [log](https://grafana.test.risingwave-cloud.xyz/explore?panes=%7B%22Y3S%22:%7B%22datasource%22:%22PE59595AED52CF917%22,%22queries%22:%5B%7B%22refId%22:%22A%22,%22expr%22:%22%7Bnamespace%3D%5C%22nexmark-bs-15-105-daily-20240220%5C%22,%20component%3D%5C%22meta%5C%22%7D%20%7C%3D%20%60too%20long%60%22,%22queryType%22:%22range%22,%22datasource%22:%7B%22type%22:%22loki%22,%22uid%22:%22PE59595AED52CF917%22%7D,%22editorMode%22:%22builder%22%7D%5D,%22range%22:%7B%22from%22:%221708444838000%22,%22to%22:%221708529618000%22%7D%7D%7D&schemaVersion=1&orgId=1) covering the full run of the test. I didn't see `duration too long` happening in this test.

**Comment by zwang28** [comment]
Another issue where compute node hangs reading block. #15239

**Comment by Xuanwo** [comment]
I can't read the logs posted in comments, so I don't know what happened yet.

Here are ideas I have:

- Enable `keep_alive` and other http settings for opendal like we do for aws s3 sdk:

https://github.com/risingwavelabs/risingwave/blob/e19aaecb1843dc9c3b309ecd8b6e863435a4fa98/src/object_store/src/object/s3.rs#L552-L567

- In some edge cases we found that http connection could hang forever, please consider enable opendal's timeout layer (or do something over the object_store trait): https://opendal.apache.org/docs/rust/opendal/layers/struct.TimeoutLayer.html

**Comment by Little-Wallace** [comment]
It confuses me that the timeout does not work when IO hang @Xuanwo 

```
        let future = async {
            self.inner
                .read(path, range)
                .verbose_instrument_await("object_store_read")
                .await
        };
        let res = match self.read_timeout.as_ref() {
            None => future.await,
            Some(read_timeout) => tokio::time::timeout(*read_timeout, future)
                .await
                .unwrap_or_else(|_| Err(ObjectError::internal("read timeout"))),
        };
```

**Comment by hzxa21** [comment]
> I can't read the logs posted in comments, so I don't know what happened yet.
> 
> Here are ideas I have:
> 
> * Enable `keep_alive` and other http settings for opendal like we do for aws s3 sdk:
> 
> https://github.com/risingwavelabs/risingwave/blob/e19aaecb1843dc9c3b309ecd8b6e863435a4fa98/src/object_store/src/object/s3.rs#L552-L567
> 
> * In some edge cases we found that http connection could hang forever, please consider enable opendal's timeout layer (or do something over the object_store trait): https://opendal.apache.org/docs/rust/opendal/layers/struct.TimeoutLayer.html

The only opendal log we saw in the testing are related to retry:
![image](https://github.com/risingwavelabs/risingwave/assets/5518566/8df952bd-3ff9-4181-ad38-0be8b662879c)


However, I don't think the retry logs are relevant to this issue because when retry indicates progress on the I/O request but what we saw in this issue is that some I/O requests are stuck forever. As mentioned by @Little-Wallace, we do enable timeout on top of each I/O call by wrapping `tokio::timeout` (this is the reason why we didn't enable TimeoutLayer in opendal) but the `tokio::timeout` didn't fire after the deadline is passed, which means there is a dead lock or blocking logic somewhere causing the timeout future not be polled.

**Comment by Li0k** [comment]
```
--- Compactor Traces ---
>> Compaction Task 50708-0
Compaction Task 50708 Split 0  [1843.496s]
  compact [!!! 1843.496s]
    compact_and_build_sst [!!! 1843.496s]
      rewind [!!! 1843.496s]
        object_store_streaming_read_read_bytes [!!! 1843.456s]

>> Compaction Task 51832-0
Compaction Task 51832 Split 0  [1062.498s]
  compact [!!! 1062.498s]
    compact_and_build_sst [!!! 1062.498s]
      add_full_key [!!! 1061.768s]
        object_store_streaming_upload_write_bytes [!!! 1061.768s]
```


Encountered Blocking of read and upload at the same time

**Comment by hzxa21** [comment]
Root cause found:
After #13618, `ObjectStoreConfig` is only passed to `MonitoredObjectStore` if the backend is `S3ObjectStore`, which causes no timeout for all other types of object store backend.

**Comment by Li0k** [comment]
can close this issue ?

**Comment by Xuanwo** [comment]
Please ping me when this happen again.
